[Xiao Yunpeng] iP

yunpeng1234 opened the pull request:
# Duke

> This works
> -Source, "Dude just trust me" :man:

Here's where to [download](https://github.com/yunpeng1234/ip/releases)

This is a bullet list
* This is **line 1**
* This is *line 2*

While this is a numbered list
1. Now this is ~~line 1~~
2. Now this is line 2

- [x] Tasks
- [ ] Not Tasks

```java
void test() {
 System.out.println("Here's some syntax highlighting");
}
```
` System.out.println("Bye")`

Thanks and have a great time.

Commit:
Add Gradle support

yunpeng1234 committed:
Level-1

yunpeng1234 committed:
Level-2

Functionality of Add, List is implemented

yunpeng1234 committed:
Level-3

Mark As Done feature is added

yunpeng1234 committed:
Level-4

Added functionality of todo, event, deadline

yunpeng1234 committed:
A - TextUiTesting

Added testing till Level - 4

yunpeng1234 committed:
Level-5

Added Exceptions to Duke

yunpeng1234 committed:
Level-5

Test Cases for Errors are added

yunpeng1234 committed:
Level-6

Added Remove functionality to Duke

yunpeng1234 committed:
A-Enums

Reworked structure to use Enums

yunpeng1234 committed:
Update READ.me

Readme is updated with usage details.

yunpeng1234 committed:
Adds Storage capabilities to Chatbot

yunpeng1234 committed:
Add functionality to parse user input as LocalDate class

yunpeng1234 committed:
Merge pull request #1 from yunpeng1234/branch-Level-7

Adds Storage capabilities to Chatbot

yunpeng1234 committed:
Merge branch 'branch-Level-8'

yunpeng1234 committed:
Remove extra print statement in code

yunpeng1234 committed:
Refactor code to extract out closely related code as classes

yunpeng1234 committed:
Organise classes into packages

yunpeng1234 committed:
Add JUnit tests to test the behaviour of the code

yunpeng1234 committed:
Add javadocs to public methods.

yunpeng1234 committed:
Add find functionality to chatbot

yunpeng1234 committed:
Tweak code to comply with coding standard.

yunpeng1234 committed:
Merge branch 'branch-A-CodingStandard'

yunpeng1234 committed:
Merge branch 'branch-Level-9'

yunpeng1234 committed:
Fix Bugs

Add Parser Class

yunpeng1234 committed:
Add JavaDocs

yunpeng1234 committed:
Update runtest.bat

yunpeng1234 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

yunpeng1234 committed:
Update Duke.java

yunpeng1234 committed:
Fix Style errors

yunpeng1234 committed:
Create basic GUI for Duke

yunpeng1234 committed:
Change MainWindow to fx:root

yunpeng1234 committed:
Fix code to remove additional code

koh-jx commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,102 @@
+        storage = new Storage(filePath);
+        try {
+            tasks = new TaskList(storage.parseFile());
+        } catch (Exception e) {
```
Pokemon exception handling (Catching them all) is generally not recommended. May I suggest listing and catching specific intended exceptions? 

eg) catch (ParseException | IllegalArgumentException | FileNotFoundException e) {...}

koh-jx commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,102 @@
+ */
+public class Duke {
+    private boolean isRunning = true;
+    private enum Commands { list, done, todo, event, deadline, delete, find }
```
Commands seem to be used in multiple files (Duke.java and Parser.java). May I suggest abstracting it to a Commands.java enum file? If not you may end up updating one without doing so for the other.

neojxuan replied:
Agreed with above. I would also recommend making Command an abstract class and the enum values children classes.

koh-jx commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    public static void main(String[] args) {
+        new Duke("./data/duke.txt").run();
```
It may be better to put the file location in the Storage class as it is directly related to storage, which in that case, you wouldn't have to pass it in into the constructor of Duke.

It can be initialised as a private static final string FILE_LOCATION, which makes it easier to find, debug and edit if necessary.

koh-jx commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,102 @@
+            } catch (DukeException e) {
+                return ui.printErrorMessage(e);
+            }
+        } catch (Exception e) {
+            System.out.println("something wrong here");
```
Another case of Pokemon Exception Handling as mentioned above, which is dangerous if you try to continue running the application, as it may lead to runtime errors.

koh-jx commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,80 @@
+        String[] splitInstructions = instruction.split(" ", 2);
+        String operative = splitInstructions[0];
+        Commands command;
+        String[] temp;
```
String[] temp is only used in 2 cases, there is no need to initialise it here.

koh-jx commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,80 @@
+     * @throws UnknownCommandError If user inputs a command that is outside the scope of the chatbot
+     */
+    public static String parse(String instruction, Ui ui, TaskList tasks) throws NoDescriptionError, UnknownCommandError {
+        String[] splitInstructions = instruction.split(" ", 2);
+        String operative = splitInstructions[0];
+        Commands command;
+        String[] temp;
+        String item;
+        String date;
+        String description;
+        String keyword;
+        Task toAdd;
+        int taskPointer;
```
Take note of the below coding standard.

![image](https://user-images.githubusercontent.com/39460640/131533697-885885e0-75d5-4fc0-9ba4-27c4772faf6b.png)

Variables should be private as well, looks like no need to keep them package-private.

koh-jx commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,110 @@
+     * @param filePath where the storage file is at.
+     */
+    public Storage (String filePath) {
+        File temp = new File(filePath);
```
temp can be declared in a smaller scope (the try-block)

koh-jx commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,110 @@
+                Task temp = parseTask(currentLine);
+                history.add(temp);
+            }
+        } catch (Exception e) {
```
There could be better exception handling here

koh-jx commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,110 @@
+     */
+    public Task parseTask(String task) {
+        String[] temp = task.split(" \\| ");
+        if (temp.length == 3) {
```
Interesting way to find for Todo. Why not do the same as the Event? (temp[0].equalsIgnoreCase("T")) It may be better as you may be adding tasks in the future whose temp.length == 3 as well.

koh-jx commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,110 @@
+            return toReturn;
+        } else {
+            Deadline toReturn = new Deadline(temp[2], temp[3]);
+            if (temp[1].equalsIgnoreCase("1")) {
```
"1" shouldn't need to IgnoreCase? I'm not too sure, correct me if I'm wrong hahaha

koh-jx commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,110 @@
+            FileWriter fileWriter = new FileWriter(file.getAbsoluteFile(), false);
+            BufferedWriter bufferedWriter = new BufferedWriter(fileWriter);
+            bufferedWriter.close();
+        } catch (Exception e) {
```
Better exception handling could be done here

koh-jx commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,110 @@
+            bufferedWriter.write(text);
+            bufferedWriter.newLine();
+            bufferedWriter.close();
+        } catch (Exception e) {
```
Same as above

koh-jx commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,60 @@
+     */
+    public Task(String description) {
+        this.description = description;
+        this.isDone = false;
```
This is interesting to note. I'm personally fine with or without.
![image](https://user-images.githubusercontent.com/39460640/131535283-c8eb7217-d265-4f07-964b-982a07a77935.png)

koh-jx commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,79 @@
+        ArrayList<Task> result = new ArrayList<>();
+        for (int i = 0; i < tasks.size(); i++) {
+            Task temp = tasks.get(i);
+            if (tasks.get(i).filterByKeyword(keyword)) {
```
Just being nitpicky here, it may be better to call temp.filterByKeyword(keyword), since you had already called tasks.get(i) earlier

koh-jx reviewed (commented):
There are a few glaring issues and there can be improvements made. Otherwise, clean and easy to understand. Comprehensive test list as well (as least more than me hahaha :P)

neojxuan commented on src/main/java/duke/MainWindow.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    /**
+     * initiatises the
```
Is this JavaDoc a WIP?

neojxuan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,80 @@
+    /**
+     * Method that parses instruction from the user.
+     *
+     * @param instruction user's instruction
```
Missing description of return value.

neojxuan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,80 @@
+ */
+public class Parser {
+    private enum Commands { list, done, todo, event, deadline, delete, find }
+
```
There seems to be an extra empty line here?

neojxuan commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,79 @@
+
+    /**
+     * Gets the number of tasks in the tasklist.
+     * @return
```
Missing description of returned value in javadoc?

neojxuan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,80 @@
+ * Class to represent parsing the commands
+ */
+public class Parser {
+    private enum Commands { list, done, todo, event, deadline, delete, find }
```
Enums contain a set of predefined constants. I think by convention, if you were to stick to using an enum for Commands, it would be something like 
```
private enum Commands { LIST, DONE, TODO, EVENT, DEADLINE, DELETE, FIND }
```

neojxuan commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,102 @@
+            if (checkBye(input)) {
+                return ui.sayFarewell();
+            }
+            try {
```
Would it be possible to avoid nested try-catch blocks?

neojxuan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,80 @@
+        default:
+            break;
+        }
+        return "";
```
Can we just return "" in the default case?

neojxuan commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,60 @@
+    public void markAsDone() {
+        this.isDone = true;
+    }
+
```
missing javadoc comment?

neojxuan commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,110 @@
+        ArrayList<Task> history = new ArrayList<Task>();
+        try {
+            FileReader fileReader = new FileReader(file);
+            BufferedReader bufferedReader = new BufferedReader(fileReader);
```
Not related to coding standard but remember to close the BufferedReader object.

neojxuan commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,102 @@
+    /**
+     * Function that checks whether the user wants to exit the program.
+     *
+     * @param instruction instruction given by the user
```
According to the module's Coding Standard, there should be punctuation behind each parameter description.
```
@param instruction Instruction given by the user.
```

neojxuan reviewed (commented):
I think your code follows the module's Coding Standard well. 😀 I mostly nitpicked on small errors like format of/missing JavaDoc comments and enum conventions. However, as the other reviewer pointed out, the scope of some variables could be changed.

yunpeng1234 committed:
Change build.gradle file to allow jar file to be built

yunpeng1234 committed:
Add variable length to chatbox

yunpeng1234 committed:
Add Assert statements to certain points in code

No asserts in current code.

It is good to check assumptions about program state so runtime can verify them.

Add some assertions in crucial parts of the program.

The program can verify it is running properly during runtime and terminate in the case of assertion failure to indicate possible bug.

yunpeng1234 committed:
Change code to have a better quality

Some parts of code were not up to t he coding quality/standards.

It makes the code harder to read and understand by others.

Change the code to suit the coding standards/practices.

This is to ensure that others have a easier time reading the code

yunpeng1234 committed:
Add streams to storage reader

JavaDocs was not clear in specify streams were used.

Storage uses streams to parse file.

JavaDoc is updated to reflect the change.

It allows others to better manage the function in the future.

yunpeng1234 committed:
Merge pull request #2 from yunpeng1234/branch-A-Assertions

Add Assert statements to certain points in code

yunpeng1234 committed:
Merge branch 'master' into branch-A-CodeQuality

yunpeng1234 committed:
Merge pull request #3 from yunpeng1234/branch-A-CodeQuality

Branch a code quality

yunpeng1234 committed:
Merge branch 'master' into branch-A-Streams

yunpeng1234 committed:
Merge pull request #4 from yunpeng1234/branch-A-Streams

Branch a streams

yunpeng1234 committed:
Add new task type of activity

yunpeng1234 committed:
Add Ci support for the future

yunpeng1234 committed:
Fix Checkstyle errors

yunpeng1234 committed:
Fix Checkstyle errors

yunpeng1234 committed:
Update doc/read.me

yunpeng1234 committed:
Fix CheckStyle errors

yunpeng1234 committed:
Set theme jekyll-theme-minimal

yunpeng1234 committed:
Set theme jekyll-theme-slate

yunpeng1234 committed:
Fix Ui bugs

yunpeng1234 committed:
Merge branch 'master' of https://github.com/yunpeng1234/ip

yunpeng1234 committed:
Fix checkstyle error

yunpeng1234 committed:
Fix CheckStyle Errors

yunpeng1234 committed:
Fix Javadoc comments
